File: Sources/Client/Sources/Game/Entity/Player.ts
import { decodeMood } from "../../../../Shared/Mood";
import { deltaTime } from "../../../Main";
import Entity from "./Entity";

// @UseRenderer(RendererFlower)
export default class Player extends Entity {
    angryT: number;
    sadT: number;

    /**
     * The player is completely removed from server (not likely death).
     */
    isRemoved: boolean;

    /**
     * Player is dev flower or not.
     */
    isDev: boolean;

    constructor(
        onlyDrawGeneralPart: boolean = false,

        id: number,
        x: number,
        y: number,
        angle: number,
        size: number,
        health: number,

        public mood: number,

        readonly nickname: string,
    ) {
        super(onlyDrawGeneralPart, id, x, y, angle, size, health);

        this.angryT = 0;
        this.sadT = 0;

        this.isRemoved = false;

        this.isDev = false;
    }

    override update() {
        super.update();

        if (this.isDead) {
            this.sadT = 1;
            this.angryT = 0;
        } else {
            const interpolationRate = deltaTime / 100;

            const { 0: isAngry, 1: isSad } = decodeMood(this.mood);

            this.angryT = Math.min(1, Math.max(0, this.angryT + (isAngry ? interpolationRate : -interpolationRate)));
            this.sadT = Math.min(1, Math.max(0, this.sadT + (!isAngry && isSad ? interpolationRate : -interpolationRate)));
        }
    }
}

File: Sources/Client/private/src/Game/UI/Shared/Entity/Player.zig
const PlayerImpl = PureRenderer.ValidatedImpl(@This());

pub const Super = Entity(PlayerImpl);

pub const Renderer = PlayerRenderingDispatcher;

/// Mood flags when player is poisoned.
/// If player is poisoned, player's mood is always locked to sad.
const poisoned_mood_flags: [2]bool = .{ false, true };

/// Mood of this player.
mood: PlayerMood.MoodBitSet = .initEmpty(),

/// Name of this player.
name: []const u8,

/// Time properties for face pupil.
angry_t: f32 = 0,
sad_t: f32 = 0,

/// Whether this player is already eliminated from server yet.
was_eliminated: bool = false,

/// Whether this player is dev flower.
is_developer: bool = false,

pub fn init(
    _: mem.Allocator,
    name: []const u8,
) PlayerImpl {
    return .{ .name = name };
}

pub fn deinit(self: *PlayerImpl, _: mem.Allocator, _: *Super) void {
    self.* = undefined;
}

pub fn update(self: *PlayerImpl, entity: *Super, delta_time: f32) void {
    if (entity.is_dead) {
        self.sad_t = 1;
        self.angry_t = 0;
    } else {
        const delta_time_100 = delta_time * comptime (1.0 / 100.0);

        const is_angry, const is_sad =
            if (entity.is_poisoned)
                poisoned_mood_flags
            else
                PlayerMood.decodeMood(self.mood);

        self.angry_t = math.clamp(self.angry_t + (if (is_angry) delta_time_100 else -delta_time_100), 0, 1);
        self.sad_t = math.clamp(self.sad_t + (if (!is_angry and is_sad) delta_time_100 else -delta_time_100), 0, 1);
    }
}

const std = @import("std");
const math = std.math;
const mem = std.mem;

const Entity = @import("Entity.zig").Entity;
const PlayerMood = @import("PlayerMood.zig");
const PureRenderer = @import("Renderers/Renderer.zig");
const PlayerRenderingDispatcher = @import("Renderers/Player/PlayerRenderingDispatcher.zig").PlayerRenderingDispatcher;
